source/common/orca/orca_parser.cc

```diff
@@ -0,0 +1,163 @@
+    orca_load_report.set_eps(value);
+  } else if (metric_name == kRpsFractionalField) {
+    orca_load_report.set_rps_fractional(value);
+  } else {
+    return absl::InvalidArgumentError(absl::StrCat("unsupported metric name: ", metric_name));
+  }
+  return absl::OkStatus();
+}
+
+absl::Status tryParseNativeHttpEncoded(const HeaderMap::GetResult& header,
+                                       OrcaLoadReport& orca_load_report) {
+  const std::vector<absl::string_view> values = parseCommaDelimitedHeader(header);
+
+  // Check for duplicate metric names here because OrcaLoadReport fields are not
+  // marked as optional and therefore don't differentiate between unset and
+  // default values.
+  absl::flat_hash_set<absl::string_view> metric_names;
+  for (const auto value : values) {
+    std::pair<absl::string_view, absl::string_view> entry =
+        absl::StrSplit(value, absl::MaxSplits(':', 1), absl::SkipWhitespace());
+    if (metric_names.contains(entry.first)) {
+      return absl::AlreadyExistsError(
+          absl::StrCat(kEndpointLoadMetricsHeader, " contains duplicate metric: ", entry.first));
+    }
+    RETURN_IF_NOT_OK(tryCopyMetricToOrcaLoadReport(entry.first, entry.second, orca_load_report));
+    metric_names.insert(entry.first);
+  }
+  return absl::OkStatus();
+}
+
+} // namespace
+
+absl::StatusOr<OrcaLoadReport> parseOrcaLoadReportHeaders(const HeaderMap& headers) {
+  OrcaLoadReport load_report;
+
+  // Binary protobuf format.
+  if (const auto header_bin = headers.get(LowerCaseString(kEndpointLoadMetricsHeaderBin));
+      !header_bin.empty()) {
+    const auto header_value = header_bin[0]->value().getStringView();
+    const std::string decoded_value = Envoy::Base64::decode(header_value);
+    if (!load_report.ParseFromString(decoded_value)) {
+      return absl::InvalidArgumentError(
+          fmt::format("unable to parse binaryheader to OrcaLoadReport: {}", header_value));
+    }
+  } else if (const auto header_native_http =
+                 headers.get(LowerCaseString(kEndpointLoadMetricsHeader));
+             !header_native_http.empty()) {
+    // Native HTTP format.
+    RETURN_IF_NOT_OK(tryParseNativeHttpEncoded(header_native_http, load_report));
+  } else if (const auto header_json = headers.get(LowerCaseString(kEndpointLoadMetricsHeaderJson));
+             !header_json.empty()) {
+    // JSON format.
+#if defined(ENVOY_ENABLE_FULL_PROTOS) && defined(ENVOY_ENABLE_YAML)
+    bool has_unknown_field = false;
+    const std::string json_string = std::string(header_json[0]->value().getStringView());
+    RETURN_IF_ERROR(
+        Envoy::MessageUtil::loadFromJsonNoThrow(json_string, load_report, has_unknown_field));
+#else
+    IS_ENVOY_BUG("JSON formatted ORCA header support not implemented for this build");
+#endif // !ENVOY_ENABLE_FULL_PROTOS || !ENVOY_ENABLE_YAML
+  } else {
+    return absl::NotFoundError("no ORCA data sent from the backend");
+  }
+
+  return load_report;
```

Comment:
As my previous comment, I will prefer the only one header's way. And let the only one header to provide different format support. And then we can register the single header as inline header.

Reply:
/wait-any

Reply:
What are your thoughts on making all three headers inline?

Additionally, what is your opinion on adding support for configuring which header to process in a later PR?

Reply:
> What are your thoughts on making all three headers inline?

inline headers is not free, so, I am afraid I won't favor it. (our an empty request headers map will take 500+bytes now)

> Additionally, what is your opinion on adding support for configuring which header to process in a later PR?

Hmmm, will it bring actual benefit to our users? I think maybe only more practices could tell the answer. But for now, in completely personal opinion, a single specific header that is clear documented and declared in some open docs, would be better.

Reply:
AFAIK gRPC already supports the "-bin" header (see [here](https://github.com/grpc/proposal/blob/master/A51-custom-backend-metrics.md#wire-format-1)). I'm not sure if it would be possible to change the contents, and it is desired that Envoy and gRPC be compatible.

Reply:
@adisuissa Got it. Thanks for point it out. gRPC is a important reference. It's fine for me to keep consistency with gRPC. Then let's start with `endpoint-load-metrics-bin` as the first supported header and format. Then we can push the PR to next step and unblock the following work of ORCA support.

My concern here is to reduce the number of new headers as much as possible. If other format/headers are required, we can treat it as an additional enhancement/extension and discuss it in independent issue/PRs. (I think plain format like text or json is useful, but it won't effect of key path of the ORCA support.)

Reply:
Thanks for sharing your thoughts and guidance. I agree that we should minimize the impact for use-cases that do not use the ORCA feature.

> My concern here is to reduce the number of new pre-defined headers as much as possible. If other format/headers are required, we can treat it as an additional enhancement/extension and discuss it in independent issue/PRs. (I think plain format like text or json is useful, but it won't effect of key path of the ORCA support.)

Trying to understand this better: is the concern introducing new "envoy defined" (non-inline) headers, or is it the performance penalty of searching for these headers?
We had an internal discussion, and wondering if adding config knobs (bootstrap), that define for an Envoy which headers are supported (plain/bin/json) addresses the issue.

Reply:
Mostly the performance and overhead. 

From my perspective, I think let's the backend services to aware which format and header (based on the front gateway's configuration) should be used is weird. Ideally, it would be better to support all different formats by the gateway. Then the backend services could select any format they like to report related stats.

So, initially, I prefer to use single header to carry different formats data. But considering the compatibility with grpc, I think it's impossible now.
So, at least let's take more consideration for others formats' design, to see which way is best for envoy and backend services.

Reply:
I agree that ORCA should require that the data plane supports all variants, which means we need to keep these limited to only a limited number. One thing I'm a little concerned about on the ORCA front is that there is no standard - my initial doc is really a proposal, and Misha also shared a proposal / design, but neither of these documents state what is required, what is optional, and so on for conformant data planes and backend services. It might be good to start such a doc, even if it's only a few lines of text that resides in Envoy `source/docs` for now.

CC @markdroth

Reply:
There is a small `Header Format` section in the [doc](https://docs.google.com/document/d/1gb_2pcNnEzTgo1EJ6w1Ol7O-EH-O_Ysu5o215N9MTAg/edit#bookmark=id.hjdltglij5g) but I'll be happy to start a doc in Envoy `source/docs`.

Reply:
Resolving.

Updated PR to include only the `endpoint-load-metrics-bin` header. We'll be putting out a follow up PR with the other formats.